src/hmem_cuda.c

```diff
@@ -484,34 +492,41 @@ static int cuda_hmem_verify_devices(void)
 	 * call to cudaGetDeviceCount() when possible.
 	 */
 
-	/* Make certain that the NVML routines are initialized */
-	nvml_ret = ofi_nvmlInit_v2();
-	if (nvml_ret != NVML_SUCCESS)
-                return -FI_ENOSYS;
+	/* Check for NVIDIA devices, if NVML library is dlopen-ed*/
+	if (cuda_attr.nvml_handle) {
 
-	/* Verify NVIDIA devices are present on the host. */
-	nvml_ret = ofi_nvmlDeviceGetCount_v2(&nvml_device_count);
-	if (nvml_ret != NVML_SUCCESS) {
-	        ofi_nvmlShutdown();
-                return -FI_ENOSYS;
-        }
+		/* Make certain that the NVML routines are initialized */
+		nvml_ret = ofi_nvmlInit_v2();
+		if (nvml_ret != NVML_SUCCESS)
+			return -FI_ENOSYS;
 
-	/* Make certain that the NVML routines get shutdown */
-        /* Note: nvmlInit / Shutdown calls are refcounted, so no harm in
-         * calling nvmlShutdown here, if the user has called nvmlInit.
-         */
-	nvml_ret = ofi_nvmlShutdown();
-	if (nvml_ret != NVML_SUCCESS)
-                return -FI_ENOSYS;
+		/* Verify NVIDIA devices are present on the host. */
+		nvml_ret = ofi_nvmlDeviceGetCount_v2(&nvml_device_count);
+		if (nvml_ret != NVML_SUCCESS) {
+			ofi_nvmlShutdown();
+			return -FI_ENOSYS;
+		}
+
+		/* Make certain that the NVML routines get shutdown */
+		/* Note: nvmlInit / Shutdown calls are refcounted, so no harm in
+		 * calling nvmlShutdown here, if the user has called nvmlInit.
+		 */
+		nvml_ret = ofi_nvmlShutdown();
+		if (nvml_ret != NVML_SUCCESS)
+			return -FI_ENOSYS;
 
-        FI_INFO(&core_prov, FI_LOG_CORE,
-                "Number of NVIDIA devices detected: %u\n",
-                nvml_device_count);
+		FI_INFO(&core_prov, FI_LOG_CORE,
+			"Number of NVIDIA devices detected: %u\n",
+			nvml_device_count);
```

Comment:
Couldn't we check nvml_device_count here and return if == 0?  We wouldn't need to initialize nvm_device_count above, and this is the only path where it might change.  Then the if statement below could be removed.

Reply:
We still need to set cuda_attr.device_count to 0 and return an error.  It seems better to have a single code path to setting the cuda_attr.device_count value and returning the error.  Otherwise, the code is duplicated and (IMHO) harder to understand.